### PR TITLE
storage/sources: Ensure we propagate kafka source errors to all source outputs

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

  * This PR fixes a recognized bug.
 Fixes https://github.com/MaterializeInc/database-issues/issues/8621 and fixes https://github.com/MaterializeInc/database-issues/issues/8691

Also reverts some of the test changes made in https://github.com/MaterializeInc/materialize/commit/3027f6c363b688e52966d3e938a8370928730a90 to match the expected correct behavior.

### Tips for reviewer

@petrosagg it's somewhat unclear to me when we should output errors to each output vs only the primary source output. For example, I've left 'connection' / ssh errors to only be reported to output 0, but now all others will be reported to each output. Would appreciate any context you have on this.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by nrainer-materialize** [comment]
Thanks for adjusting the tests as well!

**Comment by rjobanp** [comment]
I triggered nightlies on this PR which are failing for what appear to be existing issues on main. @MaterializeInc/testing if you could confirm that would be great.

**Comment by nrainer-materialize** [comment]
> I triggered nightlies on this PR which are failing for what appear to be existing issues on main. @MaterializeInc/testing if you could confirm that would be great.

I can confirm that failures are unrelated to this PR. I was unsure about the timeouts in "Checks 0dt restart" but I found a similar one in https://buildkite.com/materialize/nightly/builds/10031. I will trigger a retry of these two jobs.

File: src/storage/src/source/kafka.rs [hunk 1]
```diff
@@ -921,16 +932,19 @@ impl SourceRender for KafkaSourceConnection {
...
+                                HealthStatusMessage {
+                                    index: output.output_index,
+                                    namespace: Self::STATUS_NAMESPACE.clone(),
+                                    update: status,
+                                },
+                            );
+                        }
                     }
                     if let Some(status) = ssh_status {
+                        // We only report ssh status errors to output 0 (the primary source).
```
Comment (petrosagg) [inline src/storage/src/source/kafka.rs:947]: I would also add this to all outputs

**Review by petrosagg — APPROVED** [review]

File: src/storage/src/source/kafka.rs (hunk 1 above)
Comment (rjobanp) [reply src/storage/src/source/kafka.rs:947]: sounds good, will update everywhere we report status just to `index: 0` and output to all outputs instead

**Pull request merged**